### PR TITLE
fenicsx: update to v0.4

### PR DESCRIPTION
**Pull request by ma595**
Renamed Ufcx to FenicsUfcx and added FenicsUfcx dependency to FenicsDolfinx to remove python3 requirement. Updated to latest tagged versions.

### CODE REVIEW

File: var/spack/repos/builtin/packages/fenics-basix/package.py [hunk 1]
```diff
@@ -27,6 +29,6 @@ class FenicsBasix(CMakePackage):
 
     @property
     def root_cmakelists_dir(self):
-        if self.spec.satisfies("@main"):
+        if self.spec.satisfies("@main") or self.spec.satisfies("@0.4.0:"):
```
Comment (garth-wells) [inline var/spack/repos/builtin/packages/fenics-basix/package.py]: Is `self.spec.satisfies("@main")` still needed here?

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py [hunk 2]
```diff
@@ -15,42 +15,45 @@ class FenicsDolfinx(CMakePackage):
...
     variant("slepc", default=False, description="slepc support")
     variant("adios2", default=False, description="adios2 support")
+    variant("scotch", default=False, description="scotch support")
 
     depends_on("cmake@3.18:", type="build")
     depends_on("pkgconfig", type="build")
     depends_on("mpi")
     depends_on("hdf5+mpi")
     depends_on("boost@1.7.0:+filesystem+program_options+timer")
+    depends_on("parmetis")
```
Comment (garth-wells) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: ParMETIS could be made a variant, defaulting to `True`. What we need is that at least one of (1) ParMETIS, (2) SCOTCH or (3) KaHIP are available. Can this be easily done?

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py [hunk 3]
```diff
@@ -15,42 +15,45 @@ class FenicsDolfinx(CMakePackage):
...
+    version("0.4.1", sha256="68dcf29a26c750fcea5e02d8d58411e3b054313c3bf6fcbc1d0f08dd2851117f")
     version("0.3.0", sha256="4857d0fcb44a4e9bf9eb298ba5377abdee17a7ad0327448bdd06cce73d109bed")
     version("0.2.0", sha256="4c9b5a5c7ef33882c99299c9b4d98469fb7aa470a37a91bc5be3bb2fc5b863a4")
     version("0.1.0", sha256="0269379769b5b6d4d1864ded64402ecaea08054c2a5793c8685ea15a59af5e33")
 
     variant("kahip", default=False, description="kahip support")
-    variant("parmetis", default=False, description="parmetis support")
     variant("slepc", default=False, description="slepc support")
     variant("adios2", default=False, description="adios2 support")
+    variant("scotch", default=False, description="scotch support")
```
Comment (garth-wells) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: I think when `scotch` is activated, we need `scotch_mpi`.

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py [hunk 4]
```diff
@@ -59,15 +62,15 @@ class FenicsDolfinx(CMakePackage):
     def cmake_args(self):
         args = [
             "-DDOLFINX_SKIP_BUILD_TESTS=True",
+            "-DDOLFINX_ENABLE_PARMETIS=ON",
```
Comment (garth-wells) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: See comments above on ParMETIS.

**Review by garth-wells — CHANGES_REQUESTED** [review]
I think we can relax the `xtensor` version dependency since we can use xtensor 0.24.

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py (hunk 3 above)
Comment (ma595) [reply var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: I thought scotch mpi support would be obtained with the `depends_on("scotch+mpi", when"+scotch")` line 42.

**Comment by ma595** [comment]
> I think we can relax the `xtensor` version dependency since we can use xtensor 0.24.

It looks like xtensor is set to 0.23.10: everywhere. Are you suggesting we bump to 0.24:?

**Comment by garth-wells** [comment]
> > I think we can relax the `xtensor` version dependency since we can use xtensor 0.24.
> 
> It looks like xtensor is set to 0.23.10: everywhere. Are you suggesting we bump to 0.24:?

How about `0.23:`? Or is that a bit fragile?

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py (hunk 2 above)
Comment (ma595) [reply var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: I'm not sure how to implement this just yet. Maybe the Spack team can advise?

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: One option is through a multi-valued variant (see  https://spack.readthedocs.io/en/latest/packaging_guide.html#multi-valued-variants.)

Or, haven't tried, but you might be able to add a conflict such as:

```
    conflicts("~parmetis~scotch~kahip", msg="At least one of parmetis, scotch, and or kahip must be enabled")
````

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py [hunk 5]
```diff
@@ -59,15 +63,16 @@ class FenicsDolfinx(CMakePackage):
     def cmake_args(self):
         args = [
             "-DDOLFINX_SKIP_BUILD_TESTS=True",
-            "-DDOLFINX_ENABLE_KAHIP=%s" % (
-                'ON' if "+kahip" in self.spec else 'OFF'),
             "-DDOLFINX_ENABLE_PARMETIS=%s" % (
                 'ON' if "+parmetis" in self.spec else 'OFF'),
+            "-DDOLFINX_ENABLE_KAHIP=%s" % (
+                'ON' if "+kahip" in self.spec else 'OFF'),
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: Have you considered using any of the `CMakePackage` helper methods like `define_from_variant()`?  

For more, see https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake.

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py [hunk 6]
```diff
@@ -15,12 +15,14 @@ class FenicsDolfinx(CMakePackage):
     maintainers = ["chrisrichardson", "garth-wells", "nate-sime"]
 
     version("main", branch="main")
+    version("0.4.1", sha256="68dcf29a26c750fcea5e02d8d58411e3b054313c3bf6fcbc1d0f08dd2851117f")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py:18]: Is this deliberately `0.4.1` versus `0.4.2` with the other packages?

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/fenics-dolfinx/package.py:18]: Nvm.  Based on changes to other packages, it appears the choice is deliberate.

**Review by tldahlgren — COMMENTED** [review]
Confirmed the sha256.

File: var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py [hunk 7]
```diff
@@ -50,8 +50,8 @@ class PyFenicsDolfinx(PythonPackage):
     depends_on("py-numpy@:1.20.3", type=("build", "run"))
     depends_on("py-mpi4py", type=("build", "run"))
     depends_on("py-petsc4py", type=("build", "run"))
-    depends_on("py-pybind11@2.6.2:2.7", type=("build", "run"))
-    depends_on("xtensor@0.23:", type="build")
+    depends_on("py-pybind11", type=("build", "run"))
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py]: Are you sure you want to allow pre-`2.6.2` versions?

**Comment by tldahlgren** [comment]
Was there any intention to add the conflict or is this PR ready to go?

**Comment by garth-wells** [comment]
> Was there any intention to add the conflict or is this PR ready to go?

It's ready to go, and has been tested at https://github.com/FEniCS/dolfinx/actions/runs/2346973191.

We did some checking, and the multi-valued variant takes care of the need for at least one partitioner. Spack raises an error for an empty `partitioners=` or `partitioners=abc` if `abc` isn't in the list of  `values`, so the build isalways configured with at least one graph partitioning library.

**Review by garth-wells — APPROVED** [review]

**Review by mscroggs — APPROVED** [review]

**Comment by garth-wells** [comment]
@tldahlgren many thanks for the very helpful suggestions.

**Review by chrisrichardson — APPROVED** [review]

**Review by nate-sime — APPROVED** [review]

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**